data_structures/linked_list/doubly_linked_list.py

```diff
@@ -9,76 +9,109 @@
      Delete operation is more efficient"""
 
 
-class LinkedList:  # making main class named linked list
-    def __init__(self):
-        self.head = None
-        self.tail = None
+class LinkedList:
+    """
+    >>> linked_list = LinkedList()
+    >>> linked_list.insert_at_head("a")
+    >>> linked_list.insert_at_tail("b")
+    >>> linked_list.delete_tail()
+    'b'
+    >>> linked_list.is_empty
+    False
+    >>> linked_list.delete_head()
+    'a'
+    >>> linked_list.is_empty
+    True
+    """
 
-    def insertHead(self, x):
-        newLink = Link(x)  # Create a new link with a value attached to it
-        if self.isEmpty():  # Set the first element added to be the tail
-            self.tail = newLink
+    def __init__(self):
+        self.head = None  # First node in list
+        self.tail = None  # Last node in list
+
+    def insert_at_head(self, data):
+        new_node = Node(data)
+        if self.is_empty:
+            self.tail = new_node
+            self.head = new_node
         else:
-            self.head.previous = newLink  # newLink <-- currenthead(head)
-        newLink.next = self.head  # newLink <--> currenthead(head)
-        self.head = newLink  # newLink(head) <--> oldhead
-
-    def deleteHead(self):
-        temp = self.head
-        self.head = self.head.next  # oldHead <--> 2ndElement(head)
-        # oldHead --> 2ndElement(head) nothing pointing at it so the old head will be
-        # removed
-        self.head.previous = None
-        if self.head is None:
-            self.tail = None  # if empty linked list
-        return temp
-
-    def insertTail(self, x):
-        newLink = Link(x)
-        newLink.next = None  # currentTail(tail)    newLink -->
-        self.tail.next = newLink  # currentTail(tail) --> newLink -->
-        newLink.previous = self.tail  # currentTail(tail) <--> newLink -->
-        self.tail = newLink  # oldTail <--> newLink(tail) -->
-
-    def deleteTail(self):
-        temp = self.tail
-        self.tail = self.tail.previous  # 2ndLast(tail) <--> oldTail --> None
-        self.tail.next = None  # 2ndlast(tail) --> None
-        return temp
-
-    def delete(self, x):
+            self.head.previous = new_node
+            new_node.next = self.head
+            self.head = new_node
+
+    def delete_head(self) -> str:
+        if self.is_empty:
+            return "List is empty"
+
+        head_data = self.head.data
+        if self.head.next:
+            self.head = self.head.next
+            self.head.previous = None
+
+        else:  # If there is no next previous node
+            self.head = None
+            self.tail = None
+
+        return head_data
+
+    def insert_at_tail(self, data):
+        new_node = Node(data)
+        self.tail.next = new_node
+        new_node.previous = self.tail
+        self.tail = new_node
```

Comment:
This does not work if the list is empty. Copying your logic from `insert_at_head()`:
```suggestion
    def insert_at_tail(self, data):
        new_node = Node(data)
        if self.is_empty:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.previous = self.tail
            self.tail = new_node
```

Reply:
This was raised in #2112.